### PR TITLE
Simplify Finding Default Layout

### PR DESCRIPTION
**Pull request by apotonick**
# Summary

This PR simlifies the logic for finding default layouts in controllers when rendering.

Note: The code here is still changing, this PR was opened to have a discussion forum and to check general acceptance of this refactoring. All tests pass on my system :grimacing: 
# Implementation

While the former implementation was really hard to read and understand, the new `DefaultLayout` class cleanly provides encapsulation, and, more important, removes dozens of line of string evals, `super` calls where it was hard to know what `super` is and removes the need to hook into `inherited`.

Instead, it implements a recursion that travels up the controller inheritance chain to figure out the default layout.

Bottom line: No dynamic method defintions, no string eval, no `inherited`, and 47% less code.

### CODE REVIEW

**Comment by dmitry** [comment]
Hopefully it will be completed.

**Comment by rafaelfranca** [comment]
For what I could see this can be slower than the current implementation. The current implementation generates the method using metaprogramming at class load time. This implementation call `LayoutFinder#find` every time we are rendering a layout.

Am I missing something or this can be potentially slower?

**Comment by thedarkone** [comment]
> Am I missing something or this can be potentially slower?

@rafaelfranca yes, that is my worry/suspicion as well. I'm also not sure that this doesn't introduce unintended behavior changes. Meanwhile I'll check if the commit 83ab0a044d3b90ddb8faa9b85826f9818517a7e7 with additional tests passes on master, and then open a PR for it.

**Comment by apotonick** [comment]
We can cache the LayoutFinder at class load time. The focus here is maintainability, speeding up afterwards is simple.

@thedarkone Are you adding the test cases you found were missing to master?

**Comment by rafaelfranca** [comment]
@apotonick :+1: I just prefer to at least keep the same speed before merging. I was thinking on checking the finder at class level, but even doing this simpler cases like `layout 'my_layout`` will be slower. But I'm just wondering how much slower? Depending on the result it just doesn't matter.

**Comment by apotonick** [comment]
@rafaelfranca We can merge/fix this in the process of the AV "review", I didn't mean we have to merge it right now. Of course, we need to benchmark before merging, I totally understand this.

However, we should drop some "features" that are not even documented. While we were at this, @thedarkone and I found out that there's about 20 different ways to compute a layout and I have the objection that this blocks us from a clean refactoring. At the same time, I doubt people use all those functions anyway?

But we had that discussion already, see above..

**Comment by rafaelfranca** [comment]
> At the same time, I doubt people use all those functions anyway?

You will be surprised how much people use of Rails :smile:. I was review some of these behaviors while reviewing this patch. Some of them are very confusing and even me got it wrong as you can see at 8193a09a1f0f15c4b7624d6c24e8a779baa64d39.

I'm up to deprecate, change some of them and I think we should do what the commit revert at 8193a09a1f0f15c4b7624d6c24e8a779baa64d39 does.

I'm writing some benchmark to compare before and after the patch, let me see.

**Comment by rafaelfranca** [comment]
I used this script to benchmark before and after the patch

```ruby
require 'benchmark/ips'

require 'actionpack/abstract/layouts_test'

Benchmark.ips(5) do |bm|
  bm.report 'Layout with string' do
    Minitest.run_one_method(AbstractControllerTests::Layouts::TestBase, 'test_when_layout_is_specified_as_a_string,_render_with_that_layout')
  end

  bm.report 'Layout with proc' do
    Minitest.run_one_method(AbstractControllerTests::Layouts::TestBase, 'test_when_layout_is_specified_as_a_proc,_call_it_and_use_the_layout_returned')
  end

  bm.report 'Layout with symbol' do
    Minitest.run_one_method(AbstractControllerTests::Layouts::TestBase, 'test_when_layout_is_specified_as_a_symbol,_call_the_requested_method_and_use_the_layout_returned')
  end
end
```

### Before

```
Calculating -------------------------------------
  Layout with string   480.000  i/100ms
    Layout with proc   534.000  i/100ms
  Layout with symbol   522.000  i/100ms
-------------------------------------------------
  Layout with string      5.048k (± 8.4%) i/s -     25.440k
    Layout with proc      5.234k (± 5.7%) i/s -     26.166k
  Layout with symbol      5.045k (± 8.1%) i/s -     25.056k
```

### After

```
Calculating -------------------------------------
  Layout with string   495.000  i/100ms
    Layout with proc   498.000  i/100ms
  Layout with symbol   516.000  i/100ms
-------------------------------------------------
  Layout with string      5.063k (± 9.1%) i/s -     25.245k
    Layout with proc      5.246k (± 6.4%) i/s -     26.394k
  Layout with symbol      4.050k (±21.0%) i/s -     19.608k
```

@matthewd @thedarkone WDYT about the benchmark script?

File: actionview/lib/action_view/layouts.rb [hunk 1]
```diff
@@ -204,46 +204,15 @@ module Layouts
...
-      self._layout_conditions = {}
-      _write_layout_method
-    end
+      class_attribute :_layout_param, :_layout_conditions, :_layout_root, :instance_accessor => false
+      self._layout_param      = nil
+      self._layout_conditions = nil
+      self._layout_root       = self
 
-    delegate :_layout_conditions, to: :class
+      include LayoutMethod
```
Comment (rafaelfranca) [inline actionview/lib/action_view/layouts.rb]: There is a reason for this module? I think we can remove it and just change the implementation of `_layout` that we already have on this file.

**Comment by thedarkone** [comment]
> @thedarkone Are you adding the test cases you found were missing to master?

@apotonick yes (since they are already written).

> WDYT about the benchmark script?

@rafaelfranca it tests the whole stack, if only the layout is benched:
```ruby
require 'benchmark/ips'

require 'action_controller'
require 'action_view'

module Bench
  class Base < AbstractController::Base
    include AbstractController::Rendering
    include ActionView::Rendering
    include ActionView::Layouts

    abstract!
  end
  
  class WithString < Base
    layout 'hello'
  end
  
  class WithProc < Base
    layout proc { 'overwrite' }
  end
  
  class WithSymbol < Base
    layout :hello

    private
    def hello
      'overwrite'
    end
  end
  
  class InheritingIntermidiary < WithString; end
  class Inheriting < InheritingIntermidiary; end
end

with_string = Bench::WithString.new
with_proc   = Bench::WithProc.new
with_symbol = Bench::WithSymbol.new
inheriting  = Bench::Inheriting.new

Benchmark.ips(5) do |bm|
  bm.report 'Layout with string' do
    with_string.send(:_layout)
  end
  
  bm.report 'Layout with proc' do
    with_proc.send(:_layout)
  end
  
  bm.report 'Layout with symbol' do
    with_symbol.send(:_layout)
  end
  
  bm.report 'Layout with inheritance' do
    inheriting.send(:_layout)
  end
end
```

## Before
```
Calculating -------------------------------------
  Layout with string    99.194k i/100ms
    Layout with proc    81.674k i/100ms
  Layout with symbol    79.457k i/100ms
Layout with inheritance
                       101.416k i/100ms
-------------------------------------------------
  Layout with string      3.561M (± 3.0%) i/s -     17.855M
    Layout with proc      2.372M (± 1.7%) i/s -     11.924M
  Layout with symbol      2.047M (± 2.4%) i/s -     10.250M
Layout with inheritance
                          3.595M (± 3.3%) i/s -     17.951M
```

## After
```
Calculating -------------------------------------
  Layout with string    39.967k i/100ms
    Layout with proc    34.408k i/100ms
  Layout with symbol    34.717k i/100ms
Layout with inheritance
                        40.218k i/100ms
-------------------------------------------------
  Layout with string    620.074k (± 3.7%) i/s -      3.117M
    Layout with proc    494.184k (± 3.2%) i/s -      2.477M
  Layout with symbol    513.993k (± 1.7%) i/s -      2.569M
Layout with inheritance
                        617.246k (± 3.9%) i/s -      3.097M
```
There is some slow-down, don't know why it is not visible in your full-stack bench.

**Comment by rafaelfranca** [comment]
Hmm, the slow-down is significant.

**Comment by rafaelfranca** [comment]
Tests added by this branch were merged in https://github.com/rails/rails/commit/ce5538be68d186d414c3e21fef72b8815fc58bd6.

**Review by kingsalman88 — APPROVED** [review]
